kasa/discover.py

```diff
@@ -79,18 +79,18 @@ def datagram_received(self, data, addr) -> None:
         _LOGGER.debug("[DISCOVERY] %s << %s", ip, info)
 
         try:
-            device_class = Discover._get_device_class(info)
+            device = Discover._create_device_from_discovery_info(ip, info)
         except SmartDeviceException as ex:
             _LOGGER.debug("Unable to find device type from %s: %s", info, ex)
             return
 
-        device = device_class(ip)
-        device.update_from_discover_info(info)
-
         self.discovered_devices[ip] = device
 
-        if self.on_discovered is not None:
-            asyncio.ensure_future(self.on_discovered(device))
+        if device is not None:
+            if self.on_discovered is not None:
+                asyncio.ensure_future(self.on_discovered(device))
+        else:
+            _LOGGER.error("Received invalid response: %s", info)
```

Comment:
The noneness should be checked before adding it to `self.discovered_devices` to avoid `None`s inside the dict.

Also, in which cases this can even be none (instead of raising an exception)?

Reply:
device can't be None because the previous Discover._create_device_from_discovery_info would throw. This is to call on_discovered callback if it exists

Reply:
accepted suggestion